### DISCUSSION TITLE
Upgrading to 9.4.0 fails when applying shares

### DISCUSSION THREAD

**Discussion by keesvanbemmel** [directus/directus#10777] [Archive]
Upon first startup after upgrading to Directus 9.4.0 I'm getting this error:
10:34:15 ✨ Running migrations...
10:34:15 ✨ Applying Add Shares...
10:34:15 🚨 error alter table "directus_sessions" drop column "data" - column "data" of relation "directus_sessions" does not exist 

This will prevent our container from starting.

### To Reproduce

I did the regular upgrade steps, yarn upgrade directus and then create a snapshot of the schema. Only change in the schema was directus version.

### What version of Directus are you using?

9.4.0

### What version of Node.js are you using?

16 lts

### What database are you using?

Postgres 13

### What browser are you using?

na

### What operating system are you using?

MacOS

### How are you deploying Directus?

Docker

**Comment by joselcvarela** [comment]
Hello @keesvanbemmel 
Can you tell what was the version you had before?
I ran the migrations without any problem.

**Comment by keesvanbemmel** [comment]
9.3.0

**Comment by joselcvarela** [comment]
This is odd.
You should had `data` column on `directus_sessions`, because it was added here:
https://github.com/directus/directus/blob/5e698a77237f0ce80e4d69c7404ba21a34053bab/api/src/database/migrations/20210903A-add-auth-provider.ts#L18-L20

And it seems this was added in v9.0.0-rc95.

I think this migration ran but it failed in another place and the `data` column was removed anyways.
Although, if some migration fails, the whole process should rollback.

Do you, by any chance, have changed this table manually?

As a workaround you can simple add a `data` JSON column to `directus_sessions` in order to upgrade.

**Comment by keesvanbemmel** [comment]
Weird... I'll add it manually and will try to figure out why this data
column is missing. Can't recall doing anything manually though.

Can close this issue!

**Comment by keesvanbemmel** [comment]
I'm thinking that applying my schema snapshot before doing the migrations
somehow removed the data column (or the entire sessions table perhaps?).

**Comment by rijkvanzanten** [comment]
> I'm thinking that applying my schema snapshot before doing the migrations
somehow removed the data column (or the entire sessions table perhaps?).

Very odd! 

I'll close this for now, as it seems to be a specific issue in your particular database, rather than a migration problem in this release. Happy  to keep discussing / debugging though 🙂

**Comment by keesvanbemmel** [comment]
@rijkvanzanten still think something is wrong, albeit not with the migrations of this release. For some reason the snapshot that I have will delete directus_sessions.data see screenshot:
<img width="358" alt="CleanShot 2021-12-25 at 12 26 19@2x" src="https://user-images.githubusercontent.com/4633876/147383976-03b2ac13-7f0e-4960-844b-efb7359e1174.png">


I was under the impression that a snapshot never touches the directus_ collections? There is also no reference to directus_sessions in my schema.yaml.

Any thoughts?

MERRY CHRISTMAS ALL!!!

**Comment by rijkvanzanten** [comment]
Not sure! That being said, the snapshot should never include system collections/fields, so I something must be afoot there..

> MERRY CHRISTMAS ALL!!!

You too! 🎄 🎅🏻

**Comment by keesvanbemmel** [comment]
When applying the snapshot, this is what is returned for the data field in get-snapshot:

```js
{
    collection: 'directus_sessions',
    field: 'data',
    type: 'json',
    schema: {
      name: 'data',
      table: 'directus_sessions',
      data_type: 'json',
      default_value: null,
      generation_expression: null,
      max_length: null,
      numeric_precision: null,
      numeric_scale: null,
      is_generated: false,
      is_nullable: true,
      is_unique: false,
      is_primary_key: false,
      has_auto_increment: false,
      comment: null,
      schema: 'public',
      foreign_key_schema: null,
      foreign_key_table: null,
      foreign_key_column: null
    },
    meta: null
  }
```

The fact that meta === null will also make sure this field is not a system field (as system: true is not in this meta object).

Will try to figure out why meta === null, but not sure if I'll be able to tackle that one.

**Comment by keesvanbemmel** [comment]
Found it. Pretty sure it has something to do with removing the "data" field from sessions.yaml in this commit:
https://github.com/directus/directus/commit/dbf35a1736f8068a9bbc768cb5645d96c0777285

Not sure why it has been removed there though. Seems deliberate as it's also in the add shares migration.

@rijkvanzanten can't really assess the impact of this data field missing from that yaml file, but I thought I'd tag you here so you could assess that :)

Something to consider: our container applies schema snapshots upon startup, so the database is always up to date with the code. The order of operation is:
- schema apply (with the --yes flag, so changes aren't logged, might be smart!), this effectively removes the data column, since it's not present anymore in sessions.yaml
- Directus bootstrap, which will run migrations. Will try to drop the data column, which fails
- start directus

The first step applies our schema, but due to the error above, it completely removed the data column from our directus_sessions table. If I put it back manually, our schema with the error above will remove it again. Luckily this happened on staging :) But perhaps we can add multiple measures to make sure a schema apply does not touch any directus_ system collections or fields whatsoever, ever?

**Comment by Prochy20** [comment]
I got this error: 

```
create table `directus_shares` (`id` char(36), `name` varchar(255), `collection` varchar(64), `item` varchar(255), `role` char(36), `password` varchar(255), `user_created` char(36), `date_created` timestamp default CURRENT_TIMESTAMP, `date_start` timestamp, `date_end` timestamp, `times_used` int default '0', `max_uses` int) - ER_INVALID_DEFAULT: Invalid default value for 'date_start
``` 

in the very same situation

**Comment by keesvanbemmel** [comment]
> I got this error:
> 
> ```
> create table `directus_shares` (`id` char(36), `name` varchar(255), `collection` varchar(64), `item` varchar(255), `role` char(36), `password` varchar(255), `user_created` char(36), `date_created` timestamp default CURRENT_TIMESTAMP, `date_start` timestamp, `date_end` timestamp, `times_used` int default '0', `max_uses` int) - ER_INVALID_DEFAULT: Invalid default value for 'date_start
> ```
> 
> in the very same situation

That seems like a completely different issue. I'd create a new issue out of it if I were you. be sure to describe the exact environment (DB version etc) because it seems like a DB specific thing.

Cheers!

**Comment by Prochy20** [comment]
> > I got this error:
> > ```
> > create table `directus_shares` (`id` char(36), `name` varchar(255), `collection` varchar(64), `item` varchar(255), `role` char(36), `password` varchar(255), `user_created` char(36), `date_created` timestamp default CURRENT_TIMESTAMP, `date_start` timestamp, `date_end` timestamp, `times_used` int default '0', `max_uses` int) - ER_INVALID_DEFAULT: Invalid default value for 'date_start
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > in the very same situation
> 
> That seems like a completely different issue. I'd create a new issue out of it if I were you. be sure to describe the exact environment (DB version etc) because it seems like a DB specific thing.
> 
> Cheers!

Thanks. I have opened an issue: https://github.com/directus/directus/issues/10693

**Comment by rijkvanzanten** [comment]
The `data` column was dropped completely from the latest release @keesvanbemmel:

https://github.com/directus/directus/blob/98cf1349fdb8b0f468c00edb1714c81e2f0f90d9/api/src/database/migrations/20211211A-add-shares.ts#L22-L24

I'm a little confused why it shows up in your schema.yaml. Are you applying the schema from / to the same version? Could it be the snapshot was created on a different version of the platform?

Lets move this to a Q&A discussion, and continue there 🙂

**Reply by tchai81** [reply]
hello @rijkvanzanten,

I'm a bit worried here since our team is developing an internal application using directus and heavily relies on using the schema utility to add/remove/update existing table. (on production server as well as local dev server)

One of our team member encounters similar error on his local dev server  - some directus_sessions data column error, when applying the migration.

So my question is, what's the best practice of re-generating/re-applying the schema.yaml? Can I say that I have to regenerate the schema.yaml after every upgrade of directus version? - Schema generated on directus v9.4.0 cannot be applied on directus v9.4.2?


Thanks.

<img width="1089" alt="MicrosoftTeams-image (6)" src="https://user-images.githubusercontent.com/30285892/148522820-09932daa-ca30-4999-bcf4-30edbfd9dd78.png">

**Reply by keesvanbemmel** [reply]
If you don't mind me pitching in here, we're doing the same and I've learned a little:
1. After upgrading directus it is always a good idea to create a new snapshot. Just to be sure at least and the version number in your snapshot will match directus. Doing this will also pick up any changes if directus changes its internal field logic for example for fields in the schema.
2. If you don't have any custom migrations always run this order: migrate directus database, apply schema. This way the error/issue which is described above will not occur
3. As our snapshot is quite large, sometimes (and I don't know when exactly) the order of a snapshot is different from the order of a snapshot before when I added one field. This makes comparing it very very hard. I just opened [a PR](https://github.com/directus/directus/pull/10904) where you can optionally let the snapshot be ordered so this won't happen. Only thing is that the yaml will not be perfect (directus version for example will be somewhere halfway since it orders by key name). But I favour unpretty of incomparable :)

Hope this helps

**Reply by tchai81** [reply]
hey @keesvanbemmel 

Thanks for your reply! Noted on point 1 & point 2. Will follow the advice sequence.

My biggest challenge is to maintain a unified snapshot when working in a team,  which is relevant to your point 3:-

1. Each team member is assigned with implementing specific data model.
2. During development  they will need to create data model on their local directus server.
3.  After completing, they will need to generate the snapshot.
4. Since the schema.yaml is "shared" among developers, they will have to merge their entries manually (collection, fields & relations) to the yaml file (which is tedious),  WITHOUT altering the rest of the content (upon my request), so that it's easier to keep track on git.

My plan to mitigate this issue is to modularized schema.yaml - each data model will have it's own yaml definition (collection, fields & relations) and will be compiled to a unified one using util such as yamlinc. This ensures that only necessary changes are committed to git.

Seems like I have to scrap the above plan due to snapshot re-generation after every directus upgrade. :(

**Comment by MasonGeloso** [comment]
Hello, 
Also got the same error on two of my Directus instances when I upgraded from 9.3 to 9.4

Any fix I can put in the the interim?

![image](https://user-images.githubusercontent.com/6699213/148298650-73f7500a-bae4-48c3-83c8-c2834d9af11d.png)

**Reply by rijkvanzanten** [reply]
@MasonGeloso Did you run the database migrations?

**Reply by keesvanbemmel** [reply]
Different error. It seems like you didn't migrate your database.

**Reply by keesvanbemmel** [reply]
Haha doubling down, sorry

**Reply by MasonGeloso** [reply]
I am dumb, sorry. I migrated all but those instances. My mistake :(

**Comment by keesvanbemmel** [comment]
So in case of this error it seems like a problem with applying snapshots, custom migrations and migrations.

In my case I applied my own snapshot first. This procedure caused a drop of the data column in directus sessions since that was no longer a system column (as per directus code). I should've run the migrations first, than apply my own snapshot.

However, that order of operation will cause problems in the future (and in my past) since custom migrations might rely on the existence of stuff that is put in place by applying a snapshot.

Still not sure what the best order of operation is here, needs some thorough 2022 energy to solve :)

**Comment by viters** [comment]
I second the issue, because my workflow is exactly the same as yours @keesvanbemmel.

1. I had a running v.9.1.2 local and online.
2. I updated to 9.4.3 locally, applied migrations and regenerated schema. 
3. I pushed new schema and version to online, where it first applied schema and later run migrations.
4. It broke.

I've though about order of operations before, and as far as I am concerned - apply schema should go before migrations. Especially for rather common CMS case - create new collection and prepare initial data with migration. Although, if schema holds the same information as migrations, then this order looks impossible to achieve.

I suppose the safest solution is to run migrations and apply snapshot afterwards, and when you need a different order - you must split your task into to versions. One version: release new schema, the next version: release migrations.

There's one edge case though - new environment. You cannot use order "migrate, then apply" for freshly created environments, because it most certainly will fail. Upon this situation, you need to manually setup the initial database.
